engines/scumm/macgui/macgui_v6.h

```diff
@@ -162,22 +175,69 @@ class Octree {
 		delete node;
 	}
 
-	int getChildIndex(byte r, byte g, byte b, byte level) {
-		byte bit = (0x80 >> level);
-		byte rbit = (r & bit) >> (5 - level);
-		byte gbit = (g & bit) >> (6 - level);
-		byte bbit = (b & bit) >> (7 - level);
+	void insert(OctreeNode **node, byte r, byte g, byte b, uint level) {
+		if (*node == nullptr) {
+			*node = allocateNode(level);
+			if (level != _leafLevel) {
+				(*node)->nextNode = _reduceList[level];
+				_reduceList[level] = *node;
+			}
+		}
+
+		// Once we encounter a leaf, add the color there. This is not
+		// necessarily at the bottom of the tree, so I guess it would
+		// not be out of the question to transform the leaf into a
+		// regular node. But I saw no mention of this in the article.
 
-		return rbit | gbit | bbit;
+		if ((*node)->isLeaf) {
+			(*node)->numPixels++;
+			(*node)->sumRed += r;
+			(*node)->sumGreen += g;
+			(*node)->sumBlue += b;
+		} else {
+			byte bit = (0x80 >> level);
+			byte rbit = (r & bit) >> (5 - level);
+			byte gbit = (g & bit) >> (6 - level);
+			byte bbit = (b & bit) >> (7 - level);
+			int idx = rbit | gbit | bbit;
+
+			insert(&((*node)->child[idx]), r, g, b, level + 1);
+		}
+
+		// Usually one reduction would be enough, but it's possible
+		// that the reduction will not actually remove any leaves.
+
+		while (_numLeaves > _maxLeaves)
+			reduceTree();
 	}
 
 	void reduceTree() {
-		// TODO: Figure out why we change _leafLevel like this
-		while (!_reduceList[_leafLevel - 1])
-			_leafLevel--;
+		// In the original article, once a reduce list has been emptied
+		// the leaf level was decreased, meaning that the tree could
+		// never again grow beyond that height. I don't understand why,
+		// so I have made this a local variable instead.
+
+		int level = _leafLevel - 1;
 
-		OctreeNode *node = _reduceList[_leafLevel - 1];
-		_reduceList[_leafLevel - 1] = _reduceList[_leafLevel - 1]->nextNode;
+		while (!_reduceList[level])
+			level--;
+
+		// There are several possible approaches to picking the node to
+		// reduce. Picking the one with the largest number of pixels
+		// may leave more color for fine details. Picking the one with
+		// the smallest may will sacrifice detail, but may preserve
```

Comment:
I may be wrong, but does "may will" work here, or is it an oversight?

Reply:
Yeah, that's my error. I'll fix it later. Thanks!